### PR TITLE
evalengine: fix numeric coercibility

### LINKED ISSUE #14480: Wrong type in output with CASE and literals
For some queries where `CASE` has to be evaluated on the vtgate using the evalengine, Vitess is returning a different type than MySQL does. Here is one example of this kind of expression:

```sql
select case 
		when col = 1 then count(*) 
		else 'foo' 
	end 
from user
```

### PR DESCRIPTION
**Pull request by vmg**
## Description

@systay discovered this issue during fuzzing. We need special handling for the coercibility of numeric types in the `evalengine`, otherwise some `case` expressions will not be collated as text as you'd expect.

We've fixed the issue by replacing the generic `defaultCoercionCollation` helper with an enhanced version that takes into account the type of the value being collated. This also fixes mismatched collations for some datetime and JSON values.

Here's an example of an expression that had the right return type, but the wrong collation before this change:

```
			expression: `case column0 when 1 then column1 else column2 end`,
			values:     []sqltypes.Value{sqltypes.NewInt64(42), sqltypes.NewVarChar("sole"), sqltypes.NewInt64(0)},
			result:     `VARCHAR("0")`,
			collation:  collations.CollationUtf8mb4ID,
```

cc @dbussink 

## Related Issue(s)
Fixes #14480

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/testdata/onecase.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 [
   {
     "comment": "Add your test case here for debugging and run go test -run=One.",
-    "query": "",
+    "query": "select /*vt+ PLANNER=Gen4 */ distinct case max(tbl0.ename) when min(tbl0.job) then 'sole' else count(case when false then -27 when 'gazelle' then tbl0.deptno end) end as caggr0 from user as tbl0",
```
Comment (dbussink) [inline go/vt/vtgate/planbuilder/testdata/onecase.json:4]: @vmg I think this shouldn't be changed?

Comment (dbussink) [reply go/vt/vtgate/planbuilder/testdata/onecase.json:4]: This is also the cause of the CI failures.

File: go/vt/vtgate/semantics/dependencies.go [hunk 2]
```diff
@@ -68,6 +68,7 @@ func createUncertain(direct TableSet, recursive TableSet) *uncertain {
 		dependency: dependency{
 			direct:    direct,
 			recursive: recursive,
+			typ:       evalengine.UnknownType(),
```
Comment (vmg) [inline go/vt/vtgate/semantics/dependencies.go:71]: This was not being initialized as Unknown, so it was propagating as `NULL`.

File: go/vt/vtgate/semantics/typer.go [hunk 3]
```diff
@@ -58,11 +58,10 @@ func (t *typer) up(cursor *sqlparser.Cursor) error {
 		if !ok {
 			return nil
 		}
-		var inputType sqltypes.Type
+		inputType := sqltypes.Unknown
```
Comment (vmg) [inline go/vt/vtgate/semantics/typer.go:61]: Same here. This was not being initialized as Unknown.

**Review by systay — APPROVED** [review]

**Review by dbussink — APPROVED** [review]
Thanks for the fuzzer to find that we're not doing this right before and using the wrong coercions!

**Pull request merged**